Add support for overriding function argument types and return type

cstruct opened the pull request:
This adds support for a `override_type` and `override_return_type` attributes that can be used for manually providing a type to be used in stubs.

Usage:
```rust
#[gen_stub_pyfunction]
#[pyfunction]
#[gen_stub(override_return_type(type_repr="typing_extensions.Never", imports=("typing_extensions")))]
fn say_hello_forever<'a>(
    #[gen_stub(override_type(type_repr="collections.abc.Callable[[str]]", imports=("collections.abc")))]
    cb: Bound<'a, PyAny>,
) -> PyResult<()> {
    loop {
        cb.call1(("Hello!",))?;
    }
}
```

This fixes #241.

AndreiCravtov commented:
@cstruct does this also add the ability to override the return type? If not I might be able to contribute that some time.

cstruct commented:
No it does not as it is implemented now but I can add support for it this weekend

cstruct commented:
I've rebased on main, solved conflicts, added support for overriding return type and moved the attribute as a sub attribute to `gen_stub`.

cstruct commented:
I'll look into failed tests and clippy warnings this evening or tomorrow.

zao111222333 commented:
Great work! And it would be much better if its usage is updated in document :)
https://github.com/Jij-Inc/pyo3-stub-gen?tab=readme-ov-file#advanced-gen_stubxxx-attributes

termoshtt commented on examples/pure/src/lib.rs:
```diff
@@ -250,6 +250,48 @@ fn default_value(num: Number) -> Number {
+    }
+
+    #[getter]
+    #[gen_stub(override_type(type_repr="int", imports=()))]
```
```suggestion
    #[gen_stub(override_type(type_repr="int"))]
```

Your implementation looks to be able to skip `imports`. This case should be included in the example.

termoshtt commented on examples/pure/src/lib.rs:
```diff
@@ -250,6 +250,48 @@ fn default_value(num: Number) -> Number {
 
+#[gen_stub_pyfunction]
+#[pyfunction]
+#[gen_stub(override_type(type_repr="collections.abc.Callable[[str]]", imports=("collections.abc")))]
```
I think it's not clear at first glance that this overrides the return type. To be more explicit, I'd like to use a name like `override_return_type` instead of `override_type`.

termoshtt reviewed (changes requested):
Great work! Please check minor comments.

cstruct commented:
@termoshtt thank you for the review! I've tried to fix your comments, I've also added more contextual errors when parsing `gen_stub` attrs and fixed a merge conflict with main.

cstruct commented:
@termoshtt sorry for the churn of small fix commits, do you prefer to keep the commit history as is or do you want me to squash fix commits?

Would you be open to a PR adding https://pre-commit.com/ or something similar? I think that more people than me would benefit from automatically running `cargo fmt`, `cargo clippy`, tests, etc on commit and could possibly save you some time having to review broken code.

termoshtt commented:
> Would you be open to a PR adding [hk](https://github.com/Jij-Inc/pyo3-stub-gen/pull/hk.jdx.dev) or something similar? I think that more people than me would benefit from automatically running cargo fmt, cargo clippy, tests, etc on commit and could possibly save you some time having to review broken code.

Loosen setting. Thanks suggestion.
